### ISSUE TITLE
"make generic" broken

### ISSUE THREAD

**Issue by solardiz** [openwall/john#5135]
f2a53fa3cbae8341f02a2f76caf209ed1bf453be broke `make -f Makefile.legacy generic`. This fixes it:

```diff
+++ b/src/bench.c
@@ -382,6 +382,7 @@ char *benchmark_format(struct fmt_main *format, int salts,
 #ifdef BENCH_BUILD
        const char *where;
        if ((where = fmt_self_test(format, test_db))) {
+               static char s_error[128];
                snprintf(s_error, sizeof(s_error), "FAILED (%s)\n", where);
                return s_error;
        }
```

Found while trying to build for an arch not yet recognized by our `configure`... to fix which we need to make our `configure.ac` friendly to newer autoconf.

**Comment by solardiz** [comment]
There's also:

```
bench.c: In function 'benchmark_all':
bench.c:804: warning: unused variable 'using_int_mask'
bench.c:680: warning: unused variable 'i'
```

So maybe:

```diff
+++ b/src/bench.c
@@ -382,6 +382,7 @@ char *benchmark_format(struct fmt_main *format, int salts,
 #ifdef BENCH_BUILD
        const char *where;
        if ((where = fmt_self_test(format, test_db))) {
+               static char s_error[128];
                snprintf(s_error, sizeof(s_error), "FAILED (%s)\n", where);
                return s_error;
        }
@@ -676,7 +677,6 @@ void gather_results(struct bench_results *results)
 int benchmark_all(void)
 {
        struct fmt_main *format;
-       int i;
 #if defined(HAVE_OPENCL)
        char s_gpu[16 * MAX_GPU_DEVICES] = "";
        char s_gpu1[16 * MAX_GPU_DEVICES] = "";
@@ -685,6 +685,7 @@ int benchmark_all(void)
        const char *s_gpu1 = "";
 #endif
 #ifndef BENCH_BUILD
+       int i;
        unsigned int loop_fail = 0, loop_total = 1;
        int nvidia_mem = 0;
 #endif
@@ -799,7 +800,7 @@ AGAIN:
 #if defined(HAVE_OPENCL) || defined(HAVE_ZTEX)
                int using_int_mask = (format->params.flags & FMT_MASK) && (options.flags & FLG_MASK_CHK) &&
                        options.req_int_cand_target != 0 && mask_int_cand_target;
-#else
+#elif !defined(BENCH_BUILD)
                int using_int_mask = 0;
 #endif
 
```

**Comment by solardiz** [comment]
And there's also this:

```
wordlist.c: In function 'do_wordlist_crack':
wordlist.c:576: error: 'S_IFMT' undeclared (first use in this function)
wordlist.c:576: error: (Each undeclared identifier is reported only once
wordlist.c:576: error: for each function it appears in.)
wordlist.c:576: error: 'S_IFIFO' undeclared (first use in this function)
```

@AlekseyCherepanov You want to fix this?

**Comment by AlekseyCherepanov** [comment]
> @AlekseyCherepanov You want to fix this?

No, at least not now.

**Comment by magnumripper** [comment]
Apparently the problems in `wordlist.c` is due to us defining `_POSIX_SOURCE` (which we do for `fileno(3)`). Also defining `_BSD_SOURCE` happens to fix it but that's just from trial'n'error (supposedly only needed for `lstat()`). And that macro is deprecated.

**Comment by magnumripper** [comment]
Using `#define _XOPEN_SOURCE` instead of `_POSIX_SOURCE` also works, but I'm not feeling confident it's a correct fix on other platforms. Actually I have no idea just how to know *what* would be a canonical fix?

**Comment by magnumripper** [comment]
A non-generic build apparently sets `-D_POSIX_SOURCE -D_GNU_SOURCE -D_XOPEN_SOURCE=600` globally.
